feat:  add `array/base/take-map` (#1349)
stdlib-js/stdlib: lib/node_modules/@stdlib/array/base/take-map/README.md

distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# take-map

> Takes elements from an array, applies a mapping function using a callback, and returns a new array.

<section class="usage">

## Usage

```javascript
var takeMap = require('@stdlib/array/base/take-map');
```

### takeMap( x, indices, mode, clbk)

Combines the take and map operations into a single API call.

```javascript
var x = [ 1, 2, 3, 4 ];

function customMapping(value) {
    return value * 2;
}

var y = takeMap(x, [1, 3], 'throw', customMapping);
// returns [ 4, 8 ]
```

The function supports the following parameters:

-   **x**: input array.
-   **indices**: list of indices.
-   **mode**: index [mode][@stdlib/ndarray/base/ind].
-   **clbk**: function to apply.
-   **thisArg**: applied function execution context (optional).

</section>

<!-- /.usage -->

<section class="notes">

</section>

<!-- /.notes -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript

> var sourceArray = [10, 20, 30, 40, 50];
> var indices = [2, 0, 3, 1]; // Indices to be taken
> var mode = 'wrap'; // Index mode: 'wrap' to wrap around when exceeding array length
> 
> // Use Case 1: Multiply each value by its index
> function multiplyByIndex(value, index) {
>     return value * index;
> }
> 
> // Use Case 2: Add 100 to each value
> function addTo100(value) {
>     return value + 100;
> }

Reviewer: ```suggestion
var takeMap = require( '@stdlib/array/base/take-map' );

var sourceArray = [10, 20, 30, 40, 50];
var indices = [2, 0, 3, 1]; // Indices to be taken
var mode = 'wrap'; // Index mode: 'wrap' to wrap around when exceeding array length

// Use Case 1: Multiply each value by its index
function multiplyByIndex(value, index) {
    return value * index;
}

// Use Case 2: Add 100 to each value
function addTo100(value) {
    return value + 100;
}
```

<!--

@license Apache-2.0

Copyright (c) 2024 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# take-map

> > Take elements from an array, applies a mapping function using a callback, and returns a new array.

Reviewer 2: ```suggestion
> Take elements from an array and return a new array after applying a mapping function.
```

<section class="usage">

## Usage

```javascript

> var takeMap = require('@stdlib/array/base/take-map');

Reviewer 2: ```suggestion
var takeMap = require( '@stdlib/array/base/take-map' );
```

```

> ### takeMap( x, indices, mode, clbk)

Reviewer 2: ```suggestion
### takeMap( x, indices, mode, clbk )
```

> Combines the take and map operations into a single API call.

Reviewer 2: ```suggestion
Takes elements from an array and returns a new array after applying a mapping function.
```

```javascript

> var takeMap = require('@stdlib/array/base/take-map');

Reviewer 2: We only require the package in the first code block in the "Usage" section and then under "Examples".

```suggestion
```

var x = [ 1, 2, 3, 4 ];

function customMapping(value) {
    return value * 2;
}

> var y = takeMap(x, [1, 3], 'throw', customMapping);

Reviewer: ```suggestion
var y = takeMap(x, [ 1, 3 ], 'throw', customMapping);
```

// returns [ 4, 8 ]
```

The function supports the following parameters:

-   **x**: input array.
-   **indices**: list of indices.
-   **mode**: index [mode][@stdlib/ndarray/base/ind].
-   **clbk**: function to apply.

> 

Reviewer 2: @vr-varad Missing documentation for the `.assign` method here.

</section>

<!-- /.usage -->

<section class="notes">

</section>

<!-- /.notes -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript

> var takeMap = require('@stdlib/array/base/take-map');

Reviewer 2: It is our convention that the examples section in the `README.md` should match the examples in the `examples/index.js` file.

var sourceArray = [10, 20, 30, 40, 50];

> var indices = [2, 0, 3, 1]; // Indices to be taken

Reviewer: ```suggestion
var sourceArray = [ 10, 20, 30, 40, 50 ];
var indices = [ 2, 0, 3, 1 ]; // Indices to be taken
```

var mode = 'wrap'; // Index mode: 'wrap' to wrap around when exceeding array length

// Use Case 1: Multiply each value by its index
function multiplyByIndex(value, index) {
    return value * index;
}

// Use Case 2: Add 100 to each value
function addTo100(value) {
    return value + 100;
}

// Applying Use Case 1: Multiply each value by its index

> var multipliedResult = takeMap(sourceArray, indices, mode, multiplyByIndex);

Reviewer: ```suggestion
var multipliedResult = takeMap( sourceArray, indices, mode, multiplyByIndex );
```

> // returns <expected output>

Reviewer: ```suggestion
```

// Applying Use Case 2: Add 100 to each value

> var addedResult = takeMap(sourceArray, indices, mode, addTo100);

Reviewer: ```suggestion
var addedResult = takeMap( sourceArray, indices, mode, addTo100 );
```

> // returns <expected output>

Reviewer: ```suggestion
```

<!--

@license Apache-2.0

> Copyright (c) 2022 The Stdlib Authors.

Reviewer: ```suggestion
Copyright (c) 2024 The Stdlib Authors.
```

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

> # take

Reviewer 3: ```suggestion
# takeMap
```

> > Combine the functionality of take and map into a single API call.

Reviewer: I am not sure if this is correct description.

<section class="usage">

## Usage

```javascript

> var take = require( '@stdlib/array/base/take-map' );

Reviewer 3: The alias is `takeMap`, not `take`. Needs to be updated here and below.

```

> #### take( x, indices, mode, clbk[, thisArg])

Reviewer: ```suggestion
#### takeMap( x, indices, mode, clbk[, thisArg] )
```

> Combines the take and map operations into a single API call.

Reviewer: description needs to be updated.

Reviewer: Please update

Reviewer: cc @vr-varad

```javascript
var x = [ 1, 2, 3, 4 ];

function customMapping(value) {
    return value * 2;
}

var y = takeMap(x, [1, 3], 'throw', customMapping);
// returns [ 4, 8 ]
```

The function supports the following parameters:

-   **x**: The input array from which elements will be selected based on the provided indices.
-   **indices**: An array of indices indicating the positions of elements to be selected from the input array.
-   **mode**: index [mode][@stdlib/ndarray/base/ind].
-   **clbk**: A callback function that maps each selected value specified by indices to the corresponding output value.

> -   **thisArg**: An optional parameter that can be used to set the this value within the callback function.

Reviewer: ```suggestion
-   **x**: input array.
-   **indices**: list of indices.
-   **mode**: index [mode][@stdlib/ndarray/base/ind].
-   **clbk**: function to apply.
-   **thisArg**: applied function execution context (optional).
```

</section>

<!-- /.usage -->

<section class="notes">

</section>

<!-- /.notes -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var filledBy = require('@stdlib/array/base/filled-by');
var discreteUniform = require('@stdlib/random/base/discrete-uniform');
var linspace = require('@stdlib/array/base/linspace');
var take = require('@stdlib/array/base/take');
var takeMap = require('@stdlib/array/base/take-map');

// Generate a linearly spaced array:
var x = linspace(0, 100, 11);
console.log('Original Array:', x);

// Generate an array of random indices:
var N = discreteUniform(5, 15);
var indices = filledBy(N, discreteUniform.factory(0, x.length - 1));
console.log('Generated Indices:', indices);

// Take a random sample of elements from `x` using `take`:
var yTake = take(x, indices, 'throw');

> console.log('Taken Elements using take:', yTake);

Reviewer: ```suggestion
// returns <expected output>
```

> console.log('Taken Elements using take:', yTake);

Reviewer: Let's not use a random example here. Use explicit values.

// Define a custom mapping function:
function customMapping(value) {
    return value * 2;
}

// Take a random sample of elements from `x` and apply custom mapping using `takeMap`:
var yTakeMap = takeMap(x, indices, 'throw', customMapping);

> console.log('Taken and Mapped Elements using takeMap:', yTakeMap);

Reviewer: same comment as above.

<!--

@license Apache-2.0

Copyright (c) 2024 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

> # take-map

Reviewer 2: ```suggestion
# takeMap
```

> Take elements from an array and return a new array after applying a mapping function.

<section class="usage">

## Usage

```javascript
var takeMap = require( '@stdlib/array/base/take-map' );
```

### takeMap( x, indices, mode, clbk )

Takes elements from an array and returns a new array after applying a mapping function.

```javascript
var x = [ 1, 2, 3, 4 ];

> function customMapping(value) {

Reviewer 2: ```suggestion
function customMapping( value ) {
```

    return value * 2;
}

var y = takeMap( x, [ 1, 3 ], 'throw', customMapping );
// returns [ 4, 8 ]
```

The function supports the following parameters:

-   **x**: input array.
-   **indices**: list of indices.
-   **mode**: index [mode][@stdlib/ndarray/base/ind].
-   **clbk**: function to apply.

### takeMap.assign( x, indices, mode, out, stride, offset, clbk )

> Takes elements from an array and assigns the values to elements in a provided output array.

```javascript
var x = [ 1, 2, 3, 4 ];

var out = [ 0, 0, 0, 0, 0, 0 ];
var indices = [ 0, 0, 1, 1, 3, 3 ];

function clbk( val ) {
    return val * 2;
}

var arr = takeMap.assign( x, indices, 'throw', out, -1, out.length-1, clbk );
// returns [ 8, 8, 4, 4, 2, 2 ]

var bool = ( arr === out );
// returns true
```

The function supports the following parameters:

-   **x**: input array.
-   **indices**: list of indices.
-   **mode**: index [mode][@stdlib/ndarray/base/ind].
-   **out**: output array.
-   **stride**: output array stride.
-   **offset**: output array offset.
-   **clbk**: callback function.

> 

Reviewer 2: ```suggestion
```
